### PR TITLE
Improve error message for incompatible names in `endmacro` and `endblock`

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
To achieve this, I needed to completely change how errors were handled and to create a new type (`ParsingError`). Thanks to this, we can now add custom error messages. I'm planning to iterate on this base to provide better error messages overall and potentially even suggestion (like when the parser encounters `elif` so it can suggest to use `else if` instead).

### CODE REVIEW

File: askama_parser/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    fn from_external_error(input: I, _kind: ErrorKind, e: E) -> Self {
+        Self {
+            input,
+            message: Some(Cow::Owned(e.to_string())),
+        }
+    }
+}
+
+impl<I> ParsingError<I> {
+    pub(crate) fn from_err(error: nom::Err<Error<I>>) -> nom::Err<Self> {
```
Comment (djc) [inline askama_parser/src/error.rs:33]: Can this be a `From` or `Into` impl instead, or does that run into coherence issues?

File: askama_parser/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+use nom::error::{Error, ErrorKind, FromExternalError, ParseError};
+use std::borrow::Cow;
+
+#[derive(Debug)]
+pub(crate) struct ParsingError<I> {
```
Comment (djc) [inline askama_parser/src/error.rs:5]: We have `ParseError` already. How does this relate to that, why do we need to keep both?

File: askama_parser/src/node.rs [hunk 3]
```diff
@@ -635,6 +636,30 @@ pub struct BlockDef<'a> {
...
+                        "unexpected name `{end_name}` in `endblock` tag for unnamed `block`",
+                    ))),
+                }));
+            } else if !name.is_empty() && !end_name.is_empty() && name != end_name {
+                return Err(nom::Err::Failure(ParsingError {
+                    input: before,
+                    message: Some(Cow::Owned(format!(
+                        "expected name `{name}` in `endblock` tag, found `{end_name}`",
+                    ))),
+                }));
```
Comment (djc) [inline askama_parser/src/node.rs]: I think we can deduplicate a bunch of wrapping here?

File: testing/tests/ui/name_mismatch_endblock.stderr [hunk 4]
```diff
@@ -1,4 +1,5 @@
-error: problems parsing template source at row 1, column 27 near:
+error: error: expected name `foo` in `endblock` tag, found `not_foo`
```
Comment (djc) [inline testing/tests/ui/name_mismatch_endblock.stderr:1]: Don't love the `error: error:` here, can we avoid that?

File: askama_parser/src/node.rs [hunk 5]
```diff
@@ -626,6 +632,31 @@ impl<'a> Match<'a> {
     }
 }
 
+fn check_end_name<'a>(
```
Comment (djc) [inline askama_parser/src/node.rs:635]: Moving this should be in a separate commit from making changes to it to accomodate the second use case. IMO if you already know you're going to take it out of the parent function, better to keep it at the top-level in the introducing commit.

Please keep it ordered below the highest user.

**Review by djc — COMMENTED** [review]
Great that you're working on better errors! I like the direction, here are some nits about the implementation.

File: testing/tests/ui/name_mismatch_endblock.stderr (hunk 4 above)
Comment (GuillaumeGomez) [reply testing/tests/ui/name_mismatch_endblock.stderr:1]: Ah yes, completely missed that. I'll remove the extra `error:` I added.

File: askama_parser/src/error.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply askama_parser/src/error.rs:5]: `ParserError` is the "general" error which always contains a message and doesn't contain any `nom` information. Well, I don't see any reason why not merging both so I'll modify this type and merge them.

File: askama_parser/src/error.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply askama_parser/src/error.rs:33]: `From` forces to return `Self`, which is not the case here, hence why I didn't implement it.

File: askama_parser/src/node.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply askama_parser/src/node.rs:635]: Ok, I'll move it out directly then.

File: askama_parser/src/error.rs (hunk 1 above)
Comment (djc) [reply askama_parser/src/error.rs:33]: Ah, that makes sense.

File: askama_parser/src/error.rs (hunk 2 above)
Comment (djc) [reply askama_parser/src/error.rs:5]: If we can merge them I think that would be good thing. I think maybe we could just name it `Error`. I think it should just live in `lib.rs` -- it's too small to get its own file for my taste.

File: askama_parser/src/node.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply askama_parser/src/node.rs]: Indeed, good catch!

File: askama_parser/src/error.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply askama_parser/src/error.rs:5]: So merging both types would actually not be a great idea because of the generic argument, which have no use for users but that they would have to "live with". I can move the type to `lib.rs` though if you prefer.

**Comment by GuillaumeGomez** [comment]
Should be ready now.

File: askama_parser/src/error.rs (hunk 2 above)
Comment (djc) [reply askama_parser/src/error.rs:5]: Hmm, okay, so what's the difference between both of these types?

Comment (GuillaumeGomez) [reply askama_parser/src/error.rs:5]: The public one only (and always) contains a string and is meant to be used for displaying an error whereas the private on contains `nom` information that we need at the end to generate `ParseError`'s message.

Comment (GuillaumeGomez) [reply askama_parser/src/error.rs:5]: I'll add a comment on the type to make it clear.

**Comment by GuillaumeGomez** [comment]
I moved `ParsingError` into `lib.rs` and added documentation on it.

File: askama_parser/src/lib.rs [hunk 6]
```diff
@@ -131,6 +129,55 @@ impl fmt::Display for ParseError {
     }
 }
 
+/// This type is used to handle `nom` errors and in particular to add custom error messages.
+/// It used to generate `ParserError`.
+///
+/// It cannot be used to replace `ParseError` because it expects a generic, which would make
+/// `askama`'s users experience less good (since this generic is only needed for `nom`).
+#[derive(Debug)]
+pub(crate) struct ParsingError<I> {
```
Comment (djc) [inline askama_parser/src/lib.rs:138]: I'd like this commit squashed into the first commit (that introduces it), and I still think it should be renamed? I'm thinking maybe `ErrorContext` as I think this actually represents Askama-specific context that's spliced into nom errors in order to propagate some extra state across nom parser combinators.

File: askama_parser/src/node.rs [hunk 7]
```diff
@@ -667,6 +674,26 @@ impl<'a> BlockDef<'a> {
...
+    before: &'a str,
+    after: &'a str,
+    name: &'a str,
+    end_name: &'a str,
+    kind: &str,
+) -> IResult<&'a str, &'a str, ParsingError<&'a str>> {
+    let message = if name.is_empty() && !end_name.is_empty() {
```
Comment (djc) [inline askama_parser/src/node.rs]: I think we should move the simple case up front:

```
if name == end_name {
    return Ok((after, end_name));
}
```

Then we can simplify the subsequent `if` because we've already yeeted out the case where both `name` and `end_name` are empty.

Also, please ditch the trailing commas in single-line `format!()` calls.

File: askama_parser/src/node.rs [hunk 8]
```diff
@@ -510,6 +516,26 @@ impl<'a> Macro<'a> {
     }
 }
 
+fn check_end_name<'a>(
```
Comment (djc) [inline askama_parser/src/node.rs:519]: This shouldn't move in the second commit.

**Review by djc — COMMENTED** [review]
Some nits left, thanks for working on this!

File: askama_parser/src/lib.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply askama_parser/src/lib.rs:138]: `ErrorContext` sounds good to me.

File: askama_parser/src/node.rs (hunk 8 above)
Comment (GuillaumeGomez) [reply askama_parser/src/node.rs:519]: You told me to put it just after the highest user. Since the highest user changed, I moved it. I'm gonna revert this move then.

**Comment by GuillaumeGomez** [comment]
Updated!

File: askama_parser/src/node.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply askama_parser/src/node.rs:693]: Looks better indeed, good catch!

**Review by djc — APPROVED** [review]

File: askama_parser/src/node.rs (hunk 8 above)
Comment (djc) [reply askama_parser/src/node.rs]: You're strictly correct -- I'd have added it in the first commit in the place where it would be after the second user introduced it, balancing the concerns of (a) clean commits that don't introduce unnecessary changes within a PR, and (b) a clean ordering at the end of the PR.

**Comment by djc** [comment]
I wonder if we should bring in a type alias for `IResult<&'a str, Self, ErrorContext<&'a str>>`...

**Comment by GuillaumeGomez** [comment]
That's exactly what we're talking about in #889. :)
